### PR TITLE
Make specs for hd/1 and tl/1 more specific 

### PR DESCRIPTION
**Pull request by eksperimental**
This also stream lines with the development of the analogous functions in Elixir:

- https://github.com/elixir-lang/elixir/blob/01361af9fe582c71abdb9746c854f6fe59fbf511/lib/elixir/lib/kernel.ex#L550
- https://github.com/elixir-lang/elixir/blob/01361af9fe582c71abdb9746c854f6fe59fbf511/lib/elixir/lib/kernel.ex#L1249-L1250

### CODE REVIEW

**Comment by eksperimental** [comment]
It may require updating 
https://github.com/erlang/otp/blob/master/lib/dialyzer/src/erl_bif_types.erl
https://github.com/erlang/otp/blob/master/lib/dialyzer/src/erl_types.erl

**Comment by essen** [comment]
The input is one thing, but for the output, if you have a `term()` as a possible type, then adding more types like `[] | term()` provides no value, it'll still be just a `term()`. So I'm not sure what that part accomplishes.

File: erts/preloaded/src/erlang.erl [hunk 1]
```diff
@@ -2781,8 +2783,13 @@ term_to_iovec(_Term, _Options) ->
...
 %% Shadowed by erl_bif_types: erlang:tl/1
--spec tl(List) -> term() when
-      List :: nonempty_maybe_improper_list().
+-spec tl(List) -> Tail when
+      List :: nonempty_maybe_improper_list(T, End),
+      Tail :: maybe_improper_list(T, End) | End,
+      T :: term(),
+      End :: term().
+  
+when elem: term(), last: term()
```
Comment (essen) [inline erts/preloaded/src/erlang.erl:2792]: I don't think this line is Erlang.

Comment (eksperimental) [reply erts/preloaded/src/erlang.erl]: Oops. Removed. Thank you @essen

**Comment by eksperimental** [comment]
> The input is one thing, but for the output, if you have a `term()` as a possible type, then adding more types like `[] | term()` provides no value, it'll still be just a `term()`. So I'm not sure what that part accomplishes.

Yes. It's more for documentation purposes.  I don't know if this is idiomatic in Erlang.

File: erts/doc/src/erlang.xml [hunk 2]
```diff
@@ -2998,9 +2998,20 @@ os_prompt%</pre>
...
+       <p>Allowed in guard tests.</p>
         <p>Failure: <c>badarg</c> if <c><anno>List</anno></c> is the empty
           list <c>[]</c>.</p>
+        <pre>
+> <input>hd([]).</input>
+** exception error: bad argument
+     in function  hd/1
+        called as hd([])
+        *** argument 1: not a nonempty list
+</pre>
```
Comment (bjorng) [inline erts/doc/src/erlang.xml:3014]: I don't see the point of this example. I suggest you remove it.

File: erts/doc/src/erlang.xml [hunk 3]
```diff
@@ -11779,6 +11792,13 @@ improper_end</pre>
         <p>Allowed in guard tests.</p>
         <p>Failure: <c>badarg</c> if <c><anno>List</anno></c>
           is the empty list <c>[]</c>.</p>
+        <pre>
+> <input>tl([]).</input>
+** exception error: bad argument
+     in function  tl/1
+        called as tl([])
+        *** argument 1: not a nonempty list
+</pre>
```
Comment (bjorng) [inline erts/doc/src/erlang.xml:11801]: I don't see the point of this example. I suggest you remove it.

File: erts/preloaded/src/erlang.erl [hunk 4]
```diff
@@ -2197,8 +2197,11 @@ get_module_info(_Module, _Item) ->
     erlang:nif_error(undefined).
 
 %% Shadowed by erl_bif_types: erlang:hd/1
--spec hd(List) -> term() when
-      List :: [term(), ...].
+-spec hd(List) -> Head when
+      List :: nonempty_maybe_improper_list(T, Last),
+      Head :: term(),
+      T :: term(),
+      Last :: term().
```
Comment (bjorng) [inline erts/preloaded/src/erlang.erl:2204]: I don't find the extra variables helpful.

```suggestion
-spec hd(List) -> Head when
      List :: nonempty_maybe_improper_list(),
      Head :: term().
```

File: erts/preloaded/src/erlang.erl [hunk 5]
```diff
@@ -2781,8 +2784,11 @@ term_to_iovec(_Term, _Options) ->
     erlang:nif_error(undefined).
 
 %% Shadowed by erl_bif_types: erlang:tl/1
--spec tl(List) -> term() when
-      List :: nonempty_maybe_improper_list().
+-spec tl(List) -> Tail when
+      List :: nonempty_maybe_improper_list(T, End),
+      Tail :: maybe_improper_list(T, End) | End,
+      T :: term(),
+      End :: term().
```
Comment (bjorng) [inline erts/preloaded/src/erlang.erl:2791]: I find the extra variables and the union in the `Tail` type confusing.

```suggestion
-spec tl(List) -> Tail when
      List :: nonempty_maybe_improper_list(),
      Tail :: term().
```

**Review by bjorng — CHANGES_REQUESTED** [review]
Thanks for your pull request.

I have suggested some improvements. If you agree, please squash the commits down to one commit when you have done the changes.

File: erts/doc/src/erlang.xml (hunk 2 above)
Comment (eksperimental) [reply erts/doc/src/erlang.xml]: done

File: erts/preloaded/src/erlang.erl (hunk 4 above)
Comment (eksperimental) [reply erts/preloaded/src/erlang.erl]: done

File: erts/doc/src/erlang.xml (hunk 3 above)
Comment (eksperimental) [reply erts/doc/src/erlang.xml]: done

**Comment by eksperimental** [comment]
Please review again @bjorng .
Additionally I streamline both examples, and change "is the empty list" to "is an empty list".

**Review by bjorng — APPROVED** [review]

**Comment by bjorng** [comment]
Thanks! Add to our daily builds (along with updated preloaded modules).

**Comment by bjorng** [comment]
Thanks for your pull request.

**Pull request merged**